utils/utils.py

```diff
@@ -79,6 +82,103 @@ def options():
     return options
 
 
+def options_endswith(end):
+    def func(arg):
+        if arg.endswith(end):
+            return arg
+        raise argparse.ArgumentTypeError("value must end in '%s'" % end)
+    return func
+
+
+class ArgumentParser(argparse.ArgumentParser):
+    """
+    This lets us test for errors from argparse by overriding the error method.
+    See https://stackoverflow.com/questions/5943249
+    """
+    def _get_action_from_name(self, name):
+        """Given a name, get the Action instance registered with this parser.
+        If only it were made available in the ArgumentError object. It is
+        passed as its first arg...
+        """
+        container = self._actions
+        if name is None:
+            return None
+        for action in container:
+            if '/'.join(action.option_strings) == name:
+                return action
+            elif action.metavar == name:
+                return action
+            elif action.dest == name:
+                return action
+
+    def error(self, message):
+        exc = sys.exc_info()[1]
+        if exc:
+            exc.argument = self._get_action_from_name(exc.argument_name)
+            raise exc
+        super(ArgumentParser, self).error(message)
+
+
+def options():
+    if sys.argv[0].endswith("gather"):
+        return options_for_gather()
+    elif sys.argv[0].endswith("scan"):
+        return options_for_scan()
+
+
+def build_gather_options_parser(services):
+    parser = ArgumentParser(prefix_chars="--")
+
+    for service in services:
+        flag = "--%s" % service
+        parser.add_argument(flag, nargs=1, required=True)
+
+    parser.add_argument("--cache", action="store_true")
+    parser.add_argument("--debug", action="store_true")
+    parser.add_argument("--ignore-www", action="store_true")
+    parser.add_argument("--include-parents", action="store_true")
+    parser.add_argument("--log", nargs="+")
+    parser.add_argument("--parents", nargs="+")
+    parser.add_argument("--rdns", nargs="+")
+    parser.add_argument("--sort", action="store_true")
+    parser.add_argument("--suffix", nargs="+", required=True)
+    parser.add_argument("--timeout", nargs="+")
+    return parser
+
+
+def options_for_gather():
+    # Parse options for the ``gather`` command.
+    set_services = ("censys")
+    services = [s for s in sys.argv[1].split(",") if s not in set_services]
+    parser = build_gather_options_parser(services)
+    parsed, remaining = parser.parse_known_args()
+    for remainder in remaining:
+        if remainder.startswith("--"):
+            raise argparse.ArgumentTypeError(
```

Comment:
I guess I was expecting a little more magic from the options parser, then having to ever check for `.startswith("--")`. Is there something unusual this is trying to do to preserve the `gather` syntax?

Reply:
This check is probably being overly-cautious, but because we're dynamically creating new options here based on what gatherer is specified (for example, `dap` means that `--dap` is required), we have to account for options that the parser doesn't recognize, and this clause is trying to catch some stray flag that isn't a gatherer that the user might have entered.

It's possible that instead of dynamically turning gatherers into arguments, what we should be doing is scanning the `gatherers` directory for its contents first and constructing arguments out of that. I'm not yet sure about which of those would be better. It wouldn't be that big a change to switch over, though.

Reply:
That makes sense. Thanks for the explanation. I'm not sure which is better either, and this works for now either way.